includes/class-airtable-api.php

```diff
@@ -0,0 +1,376 @@
+<?php
+
+namespace UbcVpfoFindASpace;
+
+defined( 'ABSPATH' ) || exit;
+
+use TANIOS\Airtable\Airtable;
+
+class Airtable_Api {
+	private $van_airtable;
+	private $okan_airtable;
+
+	const CACHE_TTL      = 3600;
+	const ROOMS_PER_PAGE = 10;
+
+	private static $campus_mapping = array(
+		'vancouver' => 'van_airtable',
+		'okanagan'  => 'okan_airtable',
+	);
+
+	public function __construct( array $settings ) {
+		$this->van_airtable = new Airtable(
+			array(
+				'api_key' => $settings['api_key'],
+				'base'    => $settings['base_id_van'],
+			)
+		);
+
+		$this->okan_airtable = new Airtable(
+			array(
+				'api_key' => $settings['api_key'],
+				'base'    => $settings['base_id_okan'],
+			)
+		);
+	}
+
+	private function airtable( array $params ) {
+		$campus = sanitize_text_field( $params['campus'] );
+		return $this->{ self::$campus_mapping[ $campus ] };
+	}
+
+	private function airtable_get( string $table, array $payload, array $params ) {
+		$airtable = $this->airtable( $params );
+		$request  = $airtable->getContent( $table, $payload );
+		$response = $request->getResponse();
+
+		if ( ! $response['records'] || empty( $response['records'] ) ) {
+
+			// Check for an error from Airtable.
+			if ( isset( $response['error'] ) ) {
+				throw new \Exception(
+					'Invalid Airtable response ' .
+					wp_json_encode(
+						array(
+							'formula'  => $payload['filterByFormula'] ?? null,
+							'error'    => $response['error'],
+							'params'   => $params,
+							'response' => $response,
+							'table'    => $table,
+						)
+					)
+				);
+			}
+
+			// Return an empty result.
+			return array(
+				'records' => array(),
+				'offset'  => null,
+			);
+		}
+
+		$res = array(
+			'records' => $response['records'],
+			'offset'  => $response['offset'],
+		);
+
+		return $res;
+	}
+
+	public function get( string $func, array $params = array() ) {
+		$campus = sanitize_text_field( $params['campus'] );
+		if ( ! array_key_exists( $campus, self::$campus_mapping ) ) {
+			throw new \Exception( 'Invalid campus provided.' );
+		}
+
+		// Sort the parameters to ensure consistent cache keys.
+		ksort( $params );
+		$cache_key = null;
+
+		if ( $params['should_cache'] ) {
+			$cache_key = sprintf( '%s_%s_%s', $campus, $func, md5( wp_json_encode( $params ) ) );
+			// delete_transient( $cache_key );
+			$records = get_transient( $cache_key );
+
+			if ( $records ) {
+				return $records;
+			}
+		}
+
+		$records = call_user_func_array( array( $this, $func ), array( 'params' => $params ) );
+
+		if ( null !== $cache_key ) {
+			set_transient( $cache_key, $records, self::CACHE_TTL ); // Cache for 1 hour
```

Comment:
Please can we have some sanitization here for the item being set as a transient. Using wp_kses_post() might suffice, I'm not 100% sure what sort of data you're expecting.

Reply:
I am not sure which sanitization is best here. The content of `$records` is a JSON string from the Airtable API. I attempted to provide wp_kses_post() but that does not seem to agree with the JSON structure.

Reply:
Ah, if it's JSON, then it can be tricky, yeah. Do we know the keys in the JSON we're expecting? Likely best scenario is to decode it (which will itself determine if it's valid JSON or not) and then check for the keys we need, and then sanitize the values with the sort of data we're expecting for each. IS that feasible?

Reply:
@richardtape The JSON objects getting returned have a lot of keys (classrooms have 64, some of which are objects within objects) and it will be quite cumbersome to sanitize them each individually, particularly since different buildings and different classrooms have different levels of data completeness on the Airtable side and may not always return the same set of keys. We probably don't need every key, but we need a substantial portion of them. If you could give us an idea of anything in particular you are concerned about or want to flag, i.e. URLs, text strings, etc, that might help give us some direction in how to approach this.

Reply:
Anything that is going into the database needs to be sanitized, and when that is retrieved and output/used somewhere it needs to be escaped. If you know the types of data you're expecting as values (i.e. you know they will all be one of a url, a string of non-html text, an integer, an email address etc. then you can sanitize the data based on those expected types of value (and for all of those data types either WordPress or native PHP has functions you can use for them)). And if you know the keys are all strings, sanitize them as such. And, the final value that is going into the database should be a valid JSON object (if that's what you're storing) then validate that as such. Does that help with a direction?

Reply:
Hey @richardtape 

I have implemented some recursive sanitization for this data coming back from Airtable. Both plugins have had this same code implemented to iterate over and modify the Airtable data in place, sanitizing URLs, numeric values, and strings.

Let me know if this is adequate for database storage?

Thanks!